pkg/scheduler/internal/queue/scheduling_queue.go

```diff
@@ -336,6 +353,62 @@ func (p *PriorityQueue) Run(logger klog.Logger) {
 	}, 30*time.Second, p.stop)
 }
 
+// isPodWorthRequeuing calls QueueingHintFn of only plugins registered in pInfo.unschedulablePlugins.
+// If any QueueingHintFn returns QueueImmediately, the scheduling queue is supposed to enqueue this Pod to activeQ.
+// If no QueueingHintFn returns QueueImmediately, but some return QueueAfterBackoff,
+// the scheduling queue is supposed to enqueue this Pod to activeQ/backoffQ depending on the remaining backoff time of the Pod.
+// If all QueueingHintFn returns QueueSkip, the scheduling queue enqueues the Pod back to unschedulable Pod pool
+// because no plugin changes the scheduling result via the event.
+func (p *PriorityQueue) isPodWorthRequeuing(logger klog.Logger, pInfo *framework.QueuedPodInfo, event framework.ClusterEvent, oldObj, newObj interface{}) framework.QueueingHint {
+	if pInfo.UnschedulablePlugins.Len() == 0 {
+		logger.V(6).Info("Worth requeuing because no unschedulable plugins", "pod", klog.KObj(pInfo.Pod))
+		return framework.QueueAfterBackoff
+	}
+
+	if event.IsWildCard() {
+		logger.V(6).Info("Worth requeuing because the event is wildcard", "pod", klog.KObj(pInfo.Pod))
+		return framework.QueueAfterBackoff
+	}
+
+	hintMap, ok := p.queueingHintMap[pInfo.Pod.Spec.SchedulerName]
+	if !ok {
+		// shouldn't reach here unless bug.
+		logger.Error(nil, "No QueueingHintMap is registered for this profile", "profile", pInfo.Pod.Spec.SchedulerName, "pod", klog.KObj(pInfo.Pod))
+		return framework.QueueAfterBackoff
+	}
+
+	pod := pInfo.Pod
+	queueHint := framework.QueueSkip
+	for eventToMatch, hintfns := range hintMap {
+		if eventToMatch.Resource != event.Resource || eventToMatch.ActionType&event.ActionType == 0 {
+			continue
+		}
+
+		for _, hintfn := range hintfns {
+			if !pInfo.UnschedulablePlugins.Has(hintfn.PluginName) {
+				continue
+			}
+
+			h := hintfn.QueueingHintFn(pod, oldObj, newObj)
+			if h == framework.QueueSkip {
+				continue
+			}
+
+			if h == framework.QueueImmediately {
```

Comment:
Revisit this logic, if a pod is rejected by several plugins, and only one plugin's queueingHint returns immediately, how can we tell the pod is probably schedulable? It might be rejected by another plugin again.

Then the more reasonable logic here is - If all the unschedulable plugins return __QueueImmediately__, then we'll enqueue the pod for scheduling AFAP.

Reply:
Let me explain my take on this topic. cc @AxeZhan @ahg-g as we're talking about a similar thing in [another thread](https://github.com/kubernetes/kubernetes/pull/119155#discussion_r1260207009).

What we need is **_a clear definition_** of when to return QueueAfterBackoff and when to return QueueImmediately. Because it's currently unclear ([ref](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/types.go#L109-L121)).
And, we should consider how to handle QueueImmediately based on that.

---

First, BackoffQ is a light way of keeping throughput high by preventing pods that are "permanently unschedulable" from blocking the queue. (quote from https://github.com/kubernetes/kubernetes/pull/117561#discussion_r1189061873)

In my opinion, when to return QueueAfterBackoff and when to return QueueImmediately should be decided not by the possibility of making Pod schedulable or not, but by the reason why the Pod was in the unschedQ now.
We can split the reasons why a Pod puts back into the scheduling queue into these two:
- scheduling failure, like PodAffinity rejects Pod in Filter, NodeResourceFit rejects Pod in Filter, etc.
- non scheduling failure, like DRA needs to wait for the claim to be provisioned, DRA needs to wait for schedulingcontext to be updated by the driver etc.

So, I'm thinking that they equal when to return QueueAfterBackoff (the former) and when to return QueueImmediately (the latter). 

We should always force Pods to honor backoff if they're rejected by the scheduling failure. 
Because, in any cases, we cannot say that this Pod will 100% get scheduled in the next scheduling cycle because the cluster's situation keeps changing from moment to moment and any plugins can reject Pod in the next scheduling cycle. 

OTOH, it's OK to ignore backoff if they're not rejected by the scheduling failure. Such Pods don't have the obligation to go through backoffQ because they're not rejected by the scheduling failure.

These are my thought. And specifically speaking, DRA is the only plugin among in-tree plugins that causes Pods to be pushed back into the scheduling queue by non scheduling failure. Meaning, DRA is the only plugin that can return QueueImmediately. Other plugins like NodeAffinity we discussed are all Filter plugins, and cause Pods to be pushed back into the scheduling queue by scheduling failure. Meaning they all can return QueueAfterBackoff only.

What do you all think?

Reply:
> if DRA was NOT the only reason why a pod is unschedulable, then does it make sense to queue immediately? 

It makes sense if we follow my guidance above.
Regardless of whether DRA is the only reason or there are many other unschedulable plugins registered for a Pod as well, either way, we cannot be completely certain that a Pod will be successfully scheduled when an arrived event resolves DRA's past rejection. That depends on how each plugin is implemented and how the cluster situation gets changed, which we cannot completely manage.

So, the only thing we do is to determine whether the past rejection is due to scheduling failure or non scheduling failure, and then decide to honor backoff or skip backoff only based on that. We can skip backoff when non scheduling failure, because there is no reason that such Pods experience backoff, not because it's a higher possibility that a Pod can be schedulable in the next scheduling cycle.

> Do we have the ability to distinguish that? I suppose not.

Currently No. But, given QueueImmediately only comes from the DRA plugin among in-tree plugins at least for now, we can rethink the improvement when needed based on the usecase at that time.

Reply:
But the `DRA needs to wait for the claim to be provisioned` is also a scheduling failure and can happen together with other failure plugins. We have similar situation in volume binding when the PV isn't exist yet.

Another input: If we return QueueSkip in one failed plugin, can we tell that the pod doesn't worth requeueing, the hint should not be override by QueueAfterBackoff (I didn't consider `QueueImmediately` here).

Reply:
To be accurate, let me rephrase it that the Scheduler needs to wait for claim to be allocated/reserved. (instead of created) DRA is different from PV as you can see it has Reserve() to do that. In DRA, there is a concept WaitForFirstConsumer.
https://github.com/kubernetes/kubernetes/blob/v1.27.1/pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go#L700

That's obviously not a scheduling failure, that just waiting for the external component to do something.

> Another input: If we return QueueSkip in one failed plugin, can we tell that the pod doesn't worth requeueing, the hint should not be override by QueueAfterBackoff

No, we shouldn’t do that. We need to remember that the unschedulable plugins (set by the filter plugins) mean that all of them rejected **_some_** Nodes, doesn’t mean that all of them rejected all Nodes. So even if pluginA and pluginB are registered as unschedulable plugins, maybe we only need to solve the failure of either of them to get Pod scheduled.

Reply:
I have one proposal from the discussion here: https://github.com/kubernetes/kubernetes/pull/119517
PTAL when get a chance. 🙏